### PR TITLE
[WIP] Handle cloud-initiated provisioning

### PR DESCRIPTION
**Pull request by mlohvynenko**

### CODE REVIEW

File: iamclient/iamclient.go [hunk 1]
```diff
@@ -302,6 +307,105 @@ func (client *Client) GetCertificate(
 	return response.GetCertUrl(), response.GetKeyUrl(), nil
 }
 
+// StartProvisioning starts provisioning.
+func (client *Client) StartProvisioning(
```
Comment (al1img) [inline iamclient/iamclient.go:311]: Could fit into one line

File: iamclient/iamclient.go [hunk 2]
```diff
@@ -302,6 +307,105 @@ func (client *Client) GetCertificate(
...
+	nodeID, password string,
+) (err error) {
+	log.WithField("nodeID", nodeID).Debug("Start provisioning")
+
+	var errorInfo *cloudprotocol.ErrorInfo
```
Comment (al1img) [inline iamclient/iamclient.go:316]: Combine both vars into one var block or use assignment operator for both

File: iamclient/iamclient.go [hunk 3]
```diff
@@ -302,6 +307,105 @@ func (client *Client) GetCertificate(
...
+			CSRs:        csrs,
+		})
+		if err != nil {
+			err = aoserrors.Wrap(err)
+		}
+	}()
+
+	errorInfo = client.startProvisioning(nodeID, password)
+	if errorInfo != nil {
+		return nil
```
Comment (al1img) [inline iamclient/iamclient.go:334]: `return aoserrors.New(errorInfo.Message)`

File: iamclient/iamclient.go [hunk 4]
```diff
@@ -302,6 +307,105 @@ func (client *Client) GetCertificate(
...
+	}
+
+	csrs, errorInfo = client.createKeys(nodeID, password)
+	if errorInfo != nil {
+		return nil
```
Comment (al1img) [inline iamclient/iamclient.go:339]: ditto

File: iamclient/iamclient.go [hunk 5]
```diff
@@ -302,6 +307,105 @@ func (client *Client) GetCertificate(
...
+			ErrorInfo:   errorInfo,
+		})
+		if err != nil {
+			err = aoserrors.Wrap(err)
+		}
+	}()
+
+	errorInfo = client.applyCertificates(nodeID, csrs)
+	if errorInfo != nil {
+		return nil
```
Comment (al1img) [inline iamclient/iamclient.go:366]: ditto

File: iamclient/iamclient.go [hunk 6]
```diff
@@ -302,6 +307,105 @@ func (client *Client) GetCertificate(
...
+	}
+
+	errorInfo = client.finishProvisioning(nodeID, password)
```
Comment (al1img) [inline iamclient/iamclient.go:369]: if errorInfo != nil {
     return ...
}

File: iamclient/iamclient.go [hunk 7]
```diff
@@ -302,6 +307,105 @@ func (client *Client) GetCertificate(
...
+
+	return nil
+}
+
+// Deprovision deprovisions node.
+func (client *Client) Deprovision(
```
Comment (al1img) [inline iamclient/iamclient.go:375]: oneline

File: iamclient/iamclient.go [hunk 8]
```diff
@@ -302,6 +307,105 @@ func (client *Client) GetCertificate(
...
+
+	ctx, cancel := context.WithTimeout(context.Background(), iamRequestTimeout)
+	defer cancel()
+
+	response, err := client.provisioningService.Deprovision(
+		ctx, &pb.DeprovisionRequest{NodeId: nodeID, Password: password})
+	if err != nil {
+		errorInfo = &cloudprotocol.ErrorInfo{Message: err.Error()}
+
+		return nil
```
Comment (al1img) [inline iamclient/iamclient.go:401]: return err

File: iamclient/iamclient.go [hunk 9]
```diff
@@ -485,3 +589,104 @@ func (client *Client) subscribeNodeInfoChange() (listener pb.IAMPublicNodesServi
 
 	return listener, err
 }
+
+func (client *Client) getCertTypes(nodeID string) (
+	certTypes []string, errorInfo *cloudprotocol.ErrorInfo,
```
Comment (al1img) [inline iamclient/iamclient.go:594]: Better to return `error` rather than ErrorInfo

File: iamclient/iamclient.go [hunk 10]
```diff
@@ -302,6 +309,148 @@ func (client *Client) GetCertificate(
...
+	if errorInfo != nil {
+		return aoserrors.New(errorInfo.Message)
+	}
+
+	csrs, errorInfo = client.createKeys(nodeID, password)
+	if errorInfo != nil {
+		return aoserrors.New(errorInfo.Message)
+	}
+
+	return nil
```
Comment (mykola-kobets-epam) [inline iamclient/iamclient.go:343]: return nil overrides err from defered function. I assume "return" without arguments should be used

**Review by mykola-kobets-epam — APPROVED** [review]

File: iamclient/iamclient.go (hunk 10 above)
Comment (al1img) [reply iamclient/iamclient.go:343]: "return" without arguments is a crime. Don't do this.

**Review by al1img — APPROVED** [review]

**Pull request merged**